### PR TITLE
joyent/sdc-vmapi#40 vmapi crashes when sdc-migrate -n given name instead of UUID

### LINKED ISSUE #40: vmapi crashes when sdc-migrate -n given name instead of UUID
If you run "sdc-migrate -n somename UUID" where somename is a text string rather than a UUID of a compute node, sdc-migrate returns "curl: (52) Empty reply from server" and vmapi crashes:

```
[2019-11-01T11:24:40.656Z]  INFO: moray/19029 on 2c2752be-2211-473d-8105-5de766bbc0a6: getVmMigrations
    filter: (&(|(source_server_uuid=00000000-0000-0000-0000-0cc47aa30dba)(&(target_server_uuid=00000000-0000-0000-0000-0cc47aa30dba)(state=successful)))(vm_uuid=0cd07440-fedd-40fb-a9a2-9cba5647f500))
Uncaught AssertionError: req.params.override_server_uuid (uuid) is required

FROM
_toss (/opt/smartdc/vmapi/node_modules/assert-plus/assert.js:22:5)
Function.out.(anonymous function) [as uuid] (/opt/smartdc/vmapi/node_modules/assert-plus/assert.js:122:17)
start_create_new_record (/opt/smartdc/vmapi/lib/vm-migration/migrate.js:549:24)
Immediate.<anonymous> (/opt/smartdc/vmapi/node_modules/vasync/lib/vasync.js:875:12)
runCallback (timers.js:672:20)
tryOnImmediate (timers.js:645:5)
processImmediate [as _immediateCallback] (timers.js:617:5)
[ Nov  1 11:24:41 Stopping because all processes in service exited. ]
[ Nov  1 11:24:41 Executing stop method (:kill). ]
[ Nov  1 11:24:41 Executing start method ("/opt/smartdc/vmapi/build/node/bin/node --abort_on_uncaught_exception /opt/smartdc/vmapi/server.js &"). ]
[ Nov  1 11:24:41 Method "start" exited with status 0. ]
```

### PR DESCRIPTION
**Pull request by twhiteman**
Fixes #40

### CODE REVIEW

File: test/lib/migration.js [hunk 1]
```diff
@@ -35,6 +35,28 @@ function getJobError(job) {
...
+    assert.object(err, 'err');
+    assert.object(expectedErr, 'expectedErr');
+
+    Object.keys(expectedErr).forEach(function (prop) {
+        if (!err.hasOwnProperty(prop)) {
+            t.ok(false, 'Error object is missing the "' + prop + '" field');
+            return;
+        }
+
+        if (err[prop] != expectedErr[prop]) {
```
Comment (kusor) [inline test/lib/migration.js]:
```
if (err[prop] !== expectedErr[prop]) {
```

File: lib/vm-migration/migrate.js [hunk 2]
```diff
@@ -414,6 +414,28 @@ function migrateVm(req, res, next) {
             }
 
             if (action === 'begin' || action === 'estimate') {
+                // Validate override params.
+                if (req.params.override_server_uuid &&
```
Comment (kusor) [inline lib/vm-migration/migrate.js]: What about merging both into a single check and avoid duplicated code/wording?:
```
['override_server_uuid', 'override_uuid'].forEach(function ensureValidUuid(p) {
    if (req.params[p] && !common.validUUID(req.params[p]){
        cb(new errors.ValidationFailedError(util.format(
            'Invalid parameters: %s (%s ) is not a valid UUID', p, req.params[p]),
            [errors.invalidUuidErrorsElem(p)] ));
        return;
    }
});
```

File: test/lib/migration.js (hunk 1 above)
Comment (twhiteman) [reply test/lib/migration.js]: Will change.

File: lib/vm-migration/migrate.js (hunk 2 above)
Comment (twhiteman) [reply lib/vm-migration/migrate.js]: I tried this - but I think it ends up messier, due to the fact that the cb() is now in an inner function - so then you'd need to add a break in the inner function, add a marker that the uuid was invalid and then have a check/return in the outer function, i.e.:
```
                var hasInvalidUuid = false;
                ['override_server_uuid', 'override_uuid'].forEach(
                        function ensureValidUuid(p) {
                    if (req.params[p] && !common.validUUID(req.params[p])) {
                        cb(new errors.ValidationFailedError(util.format(
                            'Invalid parameters: %s (%s) is not a valid UUID',
                            p, req.params[p]),
                            [ errors.invalidUuidErrorsElem(
                                'override_server_uuid') ]
                        ));
                        hasInvalidUuid = true;
                        break;
                    }
                });
                if (hasInvalidUuid) {
                    return;
                }
```

This approach could also work using a for loop (no inner scope), so I'll try that and also move it into the `validateMigrationBegin` call - where it probably should have been in the first place.

**Pull request merged**